Migration to tts-wrapper and start of build script

willwade opened the pull request:
- Refactored provider_base and providers code 
- Now using tts-wrapper. Note docs on installing now just use uv and no pip install needed. It does it all for you
- started on build.py which makes a binary per platform - to be continued and made into a GH action workflow

willwade committed:
refactor to use py3-tts-wrapper

willwade committed:
adding build script

willwade committed:
attempts to include library...

willwade committed:
ruff & black lint fixes

klues commented on speech/speechManager.py:
```diff
@@ -1,89 +1,84 @@
+from provider_test_data import provider as test_provider
+
+# List of all available providers
+providers = [
```
before I've used `config.speechProviderList` in this file, now the providers are defined directly here. So I assume either `config.speechProviderList` isn't needed anymore or the providers should be defined in `config.py` as before and used here instead of the local variable `providers`?! The reason why I've moved it to `config.py` was to have everything which can be configured at a single place.

klues commented on speech/provider_test_data.py:
```diff
@@ -0,0 +1,27 @@
+from provider_base import BaseProvider
+
+
+class TestDataProvider(BaseProvider):
```
What's the difference to `eSpeakDataProvider`? It seems like the same.

klues commented on speech/start.py:
```diff
@@ -1,45 +1,56 @@
-@app.route('/speak/<text>/<providerId>/', methods=['POST', 'GET'])
-@app.route('/speak/<text>/<providerId>/<voiceId>', methods=['POST', 'GET'])
-def speak(text, providerId="", voiceId=""):
+@app.route("/speak/<text>/", methods=["POST", "GET"])
```
seems like you've used some auto-format - which IDE are you using?

willwade replied:
yes - vscode - but with ruff and black running on commit

klues commented on speech/start.py:
```diff
@@ -1,45 +1,56 @@
+@app.route("/speak/<text>/", methods=["POST", "GET"])
+@app.route("/speak/<text>/<providerId>/", methods=["POST", "GET"])
+@app.route("/speak/<text>/<providerId>/<voiceId>", methods=["POST", "GET"])
+def speak_text(text: str, providerId: str = "", voiceId: str = ""):
```
didn't know Python has TypeScript-like types :D

willwade replied:
yeah - some hate it!

klues commented on speech/speechManager.py:
```diff
@@ -1,89 +1,84 @@
+# List of all available providers
+providers = [
+    platform_provider,  # Platform-specific provider (eSpeak/AVSynth/SAPI)
+    test_provider,
```
my existing providers, like `mimic_3_data` or `msazure_data` (which you refactored) can still be used if they are imported and inserted here, correct?!

willwade replied:
ok well ive refactored again on this now.. but yes - that was the idea! my new approach is that I just need to sort mimic either out in sherpaonnx or in the tts-wrapper normally. Both possible - NB: check out Kokoro models if you like mimic https://k2-fsa.github.io/sherpa/onnx/tts/pretrained_models/kokoro.html

klues reviewed (commented):
Thanks @willwade - looks amazing at first sight!
Having the possibility to generate binary executables for different platforms is great and definitely the goal I've already had from the beginning when I've started with this. The goal is that people can use more TTS voices in AG in the easiest way possible.

It looks like you have much more experience in Python programming than I have, so probably my review is just beginner questions ;)

Did you already test these changes in combination with AsTeRICS Grid?

willwade committed:
migration step 2.

- Refactor Pt2! more use of tts-wrapper
- Test working using pytest
- Switch to UV throughout (there isnt much to depdenencies but its far nicer to work with)
- Update docs. Added swagger interface for docs
- And tons more

willwade committed:
fixing root to match original

willwade commented:
so some quick comments

- Mimic3 - I havent supported that - its pretty much dead in the water.. But I can do. I think a lot of their models we can get as onnx so I need to sort that
- Yes - ive made this all typescripty. A lot of people hate this - python was never meant to be strictly typed but it makes my ruff and black formatters happy (I have these to run on commit). 
- Ive totally rehashed big chunks of this. Probably way over the top.. Have a good look at speech_manager and start - I've added a lot of stuff to config. Now all you need to do is configure config and away you go. With sherpaonnx you get a LOT of voices with no work. Just a warning: MANY voices havent been tested. This is Meta's fault!  They released models which sound great but in eseence probably dont work 
- ive added a /docs (swagger) interface. I personally cant work an API endpoint without some pretty docs. 
- Im using uv. Its possibly not needed but it is when you do all the dev work and needing other libraries. UV IS AMAZING. its like npm for python. 

Does it work? Well you can run the uv run pytest and yay! all passing. BUT does it work with asterics? Well I get a redcross when I put in the local url to the browser. Is there something Im missing that is checking the connection? I cant see what though..

willwade committed:
fix for multiple engines - make voices list more like original

klues commented:
Thanks for your work! 👍 

I don't have the time to look at and test everything right now, but I'll do as soon as possible. But since you're much more experienced in Python than I am, I trust you that the things you've done make sense.

Mimic3 - if you say it's dead, than we don't need to support it anymore.
I just would like to keep the possibility to easily add new speech providers as before by simply adding a single python script - in case we want to test something which isn't supported by your tts-brigde.

Regarding integration to AG and the red cross: AG makes a request to `
http://localhost:5555/voices` trying to get metadata of available voices. You can test this URL by yourself and see if it returns the correct data. If yes, look at the JS console of AG and see which error messages are there.

willwade commented:
Cool. I’m working on this still. I see what’s happening with needing voices in the right structure. I’ll fix that then. 

Yes. I’ll make a couple of other changes 
1. I’ll add a technique to add mimic and other engines as they come along!
2. I’m thinking of migrating config.py to an ini file or similar. And then I’m considering making a page to configure this ini file via the flask server. 

Watch this space. I’ll mark this as draft till finished with it

klues commented:
> Yes. I’ll make a couple of other changes

both sound great - thanks! So with configuration via flask server maybe even non-tech users will be able to do configuration 🥳

willwade committed:
fix for build script

willwade committed:
getting there - now have a config screen and test interface

willwade committed:
rework around wrapper 0.9.31

willwade committed:
adding custom providers like openai

willwade committed:
fixed /speakdata endpoint - now all working. finally.

willwade committed:
false speech.ini

willwade committed:
add workflow

willwade committed:
fix build script

willwade committed:
fixes for voice endpoint and our test page

willwade committed:
tweaks for gh action build

willwade committed:
add back in portaudio

willwade committed:
cheat. add pyinstaller as its not in pyproject

should really be adding this in under dev but im lazy

willwade committed:
numpy build probs on win

willwade committed:
diif if syntax

willwade committed:
oops. idiot.

willwade committed:
upping numpy on win

willwade committed:
see if this fixes winodws

willwade committed:
maybe better syntax

willwade committed:
quotes?

willwade committed:
fix version in pyproject

willwade committed:
see if fixes imports for pyinstaller step

willwade committed:
dynanically find tts-wrapper path

willwade committed:
debug it

willwade committed:
win packages in lib

willwade committed:
marginally better readme

willwade committed:
fixed - now working in asterics grid

willwade commented:
Yay!!!

no changes needed on AG side - all working fine. I didnt realise you could add /api/ to the external provider address!


so

add http://127.0.0.1:5555/api/ 

then all works as inteneded. 


https://github.com/user-attachments/assets/4d8799a5-7c59-4ffe-8d5a-d3869d16ba40

willwade committed:
safe cleanup

willwade committed:
Add debug logging to speech endpoints

( found a problem with playht - largely because ive run out of credit but it caused a memory crash)
 Add detailed console logging for speak and speakdata endpoints
- Log URL construction and response status for better debugging
- Track speaking status updates in console
- Fix variable naming to avoid redeclaration errors
- Improve error visibility for speech service connection issues

This change helps diagnose connection and voice selection issues by providing
more detailed logging in the browser console when speech endpoints are called.

klues commented:
This looks amazing, thank you so much! 🥳
I still have to try everything on my own, but I'll definitely will merge at some point, since it greatly extends the original possibilities while at the same time making it easier to configure.

One question - in your screencast all the voices, also `sherpaonnx` are listed as "online", meaning it needs internet connection. Is this correct or is there also a possibility to run these voices using some deep learn models offline?

willwade commented:
> This looks amazing, thank you so much! 🥳 I still have to try everything on my own, but I definitely will merge at some point, since it greatly extends the original possibilities while at the same time making it easier to configure.
> 
> One question - in your screencast all the voices, also `sherpaonnx` are listed as "online", meaning it needs internet connection. Is this correct or is there also a possibility to run these voices using some deep learn model offline?

No - thats not correct. It needs to be online to download the model file for your chosen language- but once its downloaded its all offline. I need to update that!

willwade commented:
> Second question: does the bridge still cache the responses to disk (or has the possibility to do it) - in order to save API quotas?

Great catch. Its not. I need to fix that and working on that funnily enough.. It should be easy but I'm making a mess of it right now. 

> 
> Maybe a next step in evolution of this bridge would be to add a possibility to pass API keys for various services via the REST calls from AG. Then we maybe could host a running version of this somewhere on our servers and all pepole would need to do is to provide API keys within AG to use voices of a specific provider.
> 

Yeah - I was wondering about something like that. Add your speech.ini in AG and add that to a post request. It would then override the one on the server? Would that be enough? Would have to be secure.. would https post be enough?

> And the next step afterwards: add the possibility that people can "donate" their API keys (or API quotas) in some way, so that in the end people can use many voices with no or little cost. But I assume that TTS providers won't like this idea or maybe won't allow it in their terms of service.


Love the idea but like you say - may break T&C's.. 

In truth most big providers (OpenAI, Azure,  Google) the cost is ridiculously low.. ElevenLabs is actually pretty high.

willwade committed:
refactor to new abstraction method.

klues commented:
Got everything running locally - great! 🥳 

Remarks from our meeting:
* `voices` endpoint: would be great if it's returning standardized language codes like this:
   * `langCodes`: array of 2-digit language codes, e.g. `["de"]`
   * `langCodes3`: array of 3-digit language codes, e.g. `["deu"]`
   * `region`: 2 digit country code of language region, e.g. `AT` for `de-AT` (if specified by speech provider)
* it would be great to have the possibility to somehow get back a configured endpoint which the client calls itself, as an alternative to getting the binary sound data.
   * Example:
      * AG user wants to use a specific TTS service (e.g. MS Azure, Elevenlabs, ...) and provides API keys somewhere in AG
      * AG sends request to AG helper containing info like `providerId`, `apiCredentials`, `voiceId`
      * Response from AG helper: info like `requestUrl`, `responseMimeType`, `httpHeaders`
      * optional, for better privacy/security: request without API credentials, response with placeholders, which are then replaced by client
      * client does request to speech service by itself
      * advantages: less server load, no storing / passing of sensitive user data (text input) through our server

@willwade: of course this is not an assignment for you to implement these things, we're already very happy with the things you've done until now. I think the standardization of the language codes should be done before merging in order to ensure compatibility with AsTeRICS Grid, the other thing should be done somewhen else. If you want to do the language code thing, it's great, if not please bring the PR to a state where you're ok that we take over in order to complete it (in the meeting it sounded like you still would like to do some refactoring on the caching).

willwade commented:
No problem! I'll defo work with the langcodes thing. It will be configurable..

eg


```python
    # 1. BCP-47 format (default)
    voices_bcp47 = client.get_voices()
    print("\n1. BCP-47 format (default):")
    # Just print a few examples to keep output manageable
    print(json.dumps(voices_bcp47[:3], indent=2))

    # 2. ISO 639-3 format
    voices_iso = client.get_voices(langcodes="iso639_3")
    print("\n2. ISO 639-3 format:")
    print(json.dumps(voices_iso[:3], indent=2))

    # 3. Human-readable display names
    voices_display = client.get_voices(langcodes="display")
    print("\n3. Human-readable display names:")
    print(json.dumps(voices_display[:3], indent=2))

    # 4. All formats in a dictionary
    voices_all = client.get_voices(langcodes="all")
    print("\n4. All formats in a dictionary:")
    print(json.dumps(voices_all[:3], indent=2))

    # 5. Find English voices
    print("\n5. English voices (BCP-47 format):")
    english_voices = [v for v in voices_bcp47 if "en-" in v["language_codes"][0]]
    print(json.dumps(english_voices[:3], indent=2)
```

```bash
{
    "id": "fr-CA",
    "language_codes": {
      "bcp47": "fr-CA",
      "iso639_3": "fra",
      "display": "French (Canada)"
    },
    "name": "French (Canada)",
    "gender": "Unknown"
  } 
```

So the last commit - makes it all cacheable but I dont like it.. Here's why

- We do a lot of audio processing in the Grid Helper. Not great really - this is now duplcating code. But it does work..
- More dependencies in our code
- More code in general to maintain
- We lose the ability to use speak_streamed - what this method does is speak audio as soon as its being generated. Azure, Google, SherpaOnnx all can do this - it makes long utterances faster

In the big mix I figured a quicker win was to get caching working and this was my hack. But I think we can do this better but reshaping things a little. 

I'll have a think about the configured endpoint to see if I can suggest any ideas of this working

klues commented:
> No problem! I'll defo work with the langcodes thing. It will be configurable..

Great, thanks! 👍 

> We do a lot of audio processing in the Grid Helper. Not great really - this is now duplcating code. But it does work..
More dependencies in our code
More code in general to maintain

I assume you need the audio processing for playing audio within the helper?
I don't know how important the playing part actually is - and what's better - to play the audio within the helper or within AG?! Playing within AG is more flexible, since the helper can be hosted somewhere else, so maybe the focus should be in providing the audio data (or configured endpoints, see above) and the playing part of the bridge could actually be removed at some time?! I don't know, of course for playing streaming data the latency would be better, but I think it would also be possible to pass the streamed data to AG, where the stream is played within AG?!

klues commented:
> For our demo page playing is still needed but I can make this simpler.

You could also play the data with JS on the demo site. A more complex demo page, but a simpler python code ;)

> Give me a little bit of time (next week) - and I should have something

We're not in a hurry👍

deinhofer commented:
# API use by speech-Jokey application

Thanks also from my side, for this awesome tool that really helps to simplify and unify the complex task of TTS.
I would also like to use it for the Speech Jokey application (prosody parametrization for Eye-Tracking users).

But as the python-related stuff can now be outsourced to the REST API, I would like to move the rest of the Speech-Jokey application to a web-based application using speech-markdown. The plan is to adapt those two projects and integrate the API:

- [speech-markdown-vscode extension](https://github.com/speechmarkdown/speechmarkdown-vscode)
- [speech-markdown webeditor](https://github.com/speechmarkdown/speech-markdown-editor)

So from my side, the requirements regarding language code would be the same.
Additionally, it would be important to add the metadata for SSML support on a per-voice basis.

According to the [documentation of high definition voices of MS Azure](https://learn.microsoft.com/en-us/azure/ai-services/speech-service/high-definition-voices) voices can be multilingual and support a subset or a full set of ssml. This depends on the combination of ```voicename:basemodel:version``` (e.g. ```en-US-Ava:DragonHDLatestNeural```). 

So a metadata json example could be like this:
```
{
    "id": "en-US-Ava:DragonHDLatestNeural",
    "language_codes": {
      "bcp47": "fr-CA",
      "iso639_3": "fra",
      "display": "French (Canada)",
      "bcp47": "en-US",
      "iso639_3": "??",
      "display": "English (USA)",
      ... all suported language codes ...
    },
    "ssml_support" : "full|subset|no",
    "name": "English (USA) Ava",
    "gender": "female"
  }
```
As Amazong Polly, MS and others do have an SSML flavour, it could also make sense to define the SSML flavour.

This would be great, but should probably also be put into another PR. I also have an apprentice until July who could help in adding metadata.

willwade committed:
upgrading to new tts-wrapper code. nb: beta tagged release. better lang support

willwade commented:
Ok some big changes in this

- MMS/SherpaOnnx models now in python lib - so you shouldnt see them (it was a bit of a bug)
- Voices now return a mix of langcodes as suggested above.
- Sorted out a bunch of issues 
- Simplifed init of engines - now just one import per engine rather than two
- We actually can now support openai in tts-wrapper - but left our existing code as a demo really
- Made a few more small tweaks from improving the wrapper 

needs some thorough testing though - ive done a lot of this with minimal testing to reduce overhead on my keys!

willwade commented:
> But as the python-related stuff can now be outsourced to the REST API, I would like to move the rest of the Speech-Jokey application to a web-based application using speech-markdown. The plan is to adapt those two projects and integrate the API:

Cool. I love speech markdown - but I'm a little concerned  as last time I looked at it its pretty out of date. Maybe I'm wrong though? 

> According to the [documentation of high definition voices of MS Azure](https://learn.microsoft.com/en-us/azure/ai-services/speech-service/high-definition-voices) voices can be multilingual and support a subset or a full set of ssml. This depends on the combination of voicename:basemodel:version (e.g. en-US-Ava:DragonHDLatestNeural).

yeah so our get_voices should support this - from memory our azure engine does give a load of metadata but the voiceid = en-US-Ava:DragonHDLatestNeural layout


See https://willwade.github.io/tts-wrapper/docs/guides/voices for a overview of our data coming back

> As Amazong Polly, MS and others do have an SSML flavour, it could also make sense to define the SSML flavour.

for sure - speechmd does support polly too

> Styles (e.g. cheerful, excited, angry,... and roles could also be interesting, see https://learn.microsoft.com/nb-no/azure/ai-services/speech-service/language-support?tabs=tts#voice-styles-and-roles


For sure - although openai mess this up. They dont use SSML but a prompt on voice generation. I considered mapping ssml azure styles to make a prompt but this is messy because with SSML you can chunk a utterance with multiple styles. 


So in short - our stuff supports SSML as much as we can - if you create the SSML - just pass a speak() or speak_streamed - text or ssml.

deinhofer commented:
> Ok some big changes in this
> 
>     * MMS/SherpaOnnx models now in python lib - so you shouldnt see them (it was a bit of a bug)
> 
>     * Voices now return a mix of langcodes as suggested above.
> 
>     * Sorted out a bunch of issues
> 
>     * Simplifed init of engines - now just one import per engine rather than two
> 
>     * We actually can now support openai in tts-wrapper - but left our existing code as a demo really
> 
>     * Made a few more small tweaks from improving the wrapper
> 
> 
> needs some thorough testing though - ive done a lot of this with minimal testing to reduce overhead on my keys!

Thanks @willwade 

My student would do the intensive testing now. We already have found an issue that the amazon polly voice is not listed altough configured.
How should we provide bug reports? Could you enable issue tracking in your [fork](https://github.com/willwade/AsTeRICS-Grid-Helper/tree/main)?
Then we would add all issues found, ok?
